### PR TITLE
LSP configuration via didChangeConfiguration

### PR DESCRIPTION
**Pull request by danielwelch**
Addressing #1443

>Disclaimer: This is my first time writing vimscript. Feedback is extremely welcome. I'm sure I missed the boat on something here but I'm happy to help any way I can with some guidance.

Adding the ability to configure LSPs via the [`workspace/didChangeConfiguration`](https://microsoft.github.io/language-server-protocol/specification#workspace_didChangeWorkspaceFolders) method. I've added the `lsp_config` key to linter definition, added a new `ale#lsp#message#DidChangeConfiguration`, and added a call to `ale#lsp#Send` this message during `ale#lsp#StartLSP` if `lsp_config` is set.

This leaves it up to each LSP linter to define its own configuration variable and pass that to the linter definition. I've added an example implementation for `pyls`.

### CODE REVIEW

**Comment by danielwelch** [comment]
Are we sure these are the same thing? I'm most familiar with the implementation of RLS, so I'll use that as an example... Here's the [`initialization_options`](https://github.com/rust-lang-nursery/rls/blob/f92b4e8db8c9e4cfa032972053ce6491fb25712f/src/lsp_data.rs#L263) available to RLS and [here are the configuration options](https://github.com/rust-lang-nursery/rls/blob/master/src/config.rs#L118) available via `workspace/didChangeConfiguration`.

**Comment by w0rp** [comment]
I guess they are different. For `pyls`, the options people will really care about can all be given in `initialization_options`, like disabling or enabling particular plugins.

**Comment by w0rp** [comment]
Add support for `didChangeConfiguration`, but take away your changes for `pyls` for now.

**Comment by danielwelch** [comment]
I removed the pyls changes. I'll probably add the RLS changes in a separate pull request once this is merged.

File: autoload/ale/lsp_linter.vim [hunk 1]
```diff
@@ -180,6 +180,14 @@ function! ale#lsp_linter#StartLSP(buffer, linter) abort
 
     let l:language_id = ale#util#GetFunction(a:linter.language_callback)(a:buffer)
 
+    if has_key(a:linter, 'lsp_config')
```
Comment (w0rp) [inline autoload/ale/lsp_linter.vim]: You can use `if !empty(get(a:linter, 'lsp_config'))` here.

You'll have to update `ale#linter#PreProcess` to accept `lsp_config` as an option, and check the type, and update the tests to check that it's allowed and that it appears in the output okay.

**Comment by tmc** [comment]
would love to see this land.

File: autoload/ale/lsp_linter.vim (hunk 1 above)
Comment (w0rp) [reply autoload/ale/lsp_linter.vim]: Look at what the function does. The `ale#Define` options are validated and new Dictionary values are created from the options. That function hasn't been modified, so this option doesn't work yet.

File: autoload/ale/linter.vim [hunk 2]
```diff
@@ -255,6 +255,13 @@ function! ale#linter#PreProcess(filetype, linter) abort
         elseif has_key(a:linter, 'initialization_options')
             let l:obj.initialization_options = a:linter.initialization_options
         endif
+
+        if has_key(a:linter, 'lsp_config')
+            if type(a:linter.lsp_config) isnot v:t_dict
+                throw '`lsp_config` must be a Dictionary'
+            endif
+        endif
```
Comment (w0rp) [inline autoload/ale/linter.vim]: You need to copy the dictionary to `l:obj`, and add tests for the new option.

Comment (danielwelch) [reply autoload/ale/linter.vim]: Thanks for the feedback. I'll work on tests when I get the chance.

**Comment by tmc** [comment]
Great to see progress here!

**Comment by tmc** [comment]
What is missing here, tests?

**Comment by RyanSquared** [comment]
Some tests that make sure that bad input is thrown would be nice, I think. You should always test every possible branch, *especially* the ones you want to fail.

**Comment by danielwelch** [comment]
Yep, tests are still missing. Specifically the PreProcess tests still need to be written.

**Comment by danielwelch** [comment]
@w0rp added tests to check PreProcess (to check that `lsp_config` is accepted and processed appropriately, and that it throws if `lsp_config` is not a dictionary)

**Comment by w0rp** [comment]
Cheers! :beers:

**Comment by jparise** [comment]
Would this also benefit from an `lsp_config_callback` variant so that `lsp_config` doesn't need to be a static dictionary and to support `ale#VarFunc`?

**Pull request merged**